File: py/DREAM/Settings/TimeStepper.py
#
# TimeStepper settings object
# #################################

import numpy as np

from .. DREAMException import DREAMException
from . ToleranceSettings import ToleranceSettings


TYPE_CONSTANT = 1
TYPE_ADAPTIVE = 2
TYPE_IONIZATION = 3


class TimeStepper:
    
    def __init__(self, ttype=1, checkevery=0, tmax=None, dt=None, nt=None, nSaveSteps=0, reltol=1e-2, verbose=False, constantstep=False, terminatefunc=None):
        """
        Constructor.
        """
        self.set(ttype=ttype, checkevery=checkevery, tmax=tmax, dt=dt, nt=nt, nSaveSteps=nSaveSteps, reltol=reltol, verbose=verbose, constantstep=constantstep, terminatefunc=None)
        

    def set(self, ttype=1, checkevery=0, tmax=None, dt=None, nt=None, nSaveSteps=0, reltol=1e-2, verbose=False, constantstep=False, minsavedt=0, terminatefunc=None):
        """
        Set properties of the time stepper.
        """
        self.type = int(ttype)

        self.setCheckInterval(checkevery)
        self.setTmax(tmax)
        self.setDt(dt)
        self.setNt(nt)
        self.setMinSaveTimestep(minsavedt)
        self.setNumberOfSaveSteps(nSaveSteps)
        self.setVerbose(verbose)
        self.setConstantStep(constantstep)       
        self.tolerance = ToleranceSettings()
        self.tolerance.set(reltol=reltol)
        self.terminatefunc = terminatefunc
        
        self.dtmax = None
        self.automaticstep = None
        self.safetyfactor = None


    def __contains__(self, item):
        return (item in self.todict(False))


    def __getitem__(self, key):
        return self.todict(False)[key]


    ######################
    # SETTERS
    ######################
    def setCheckInterval(self, checkevery):
        if checkevery < 0:
            raise DREAMException("TimeStepper: Invalid value assigned to 'checkevery': {}".format(checkevery))
        
        self.checkevery = int(checkevery)


    def setConstantStep(self, constantstep):
        self.constantstep = bool(constantstep)


    def setDt(self, dt):
        if dt is None:
            self.dt = None
            return

        if dt < 0 or (dt == 0 and self.type != TYPE_IONIZATION):
            raise DREAMException("TimeStepper: Invalid value assigned to 'dt': {}".format(tmax))
        if self.nt is not None and dt > 0:
            raise DREAMException("TimeStepper: 'dt' may not be set alongside 'nt'.")
            
        self.dt = float(dt)


    def setMinSaveTimestep(self, dt):
        """
        For the adapative ionization-based time stepper, sets the minimum
        time which must elapse between two saved time steps.
        """
        self.minsavedt = dt


    def setNt(self, nt):
        if nt is None:
            self.nt = None
            return

        if nt <= 0:
            raise DREAMException("TimeStepper: Invalid value assigned to 'dt': {}".format(tmax))
        if self.dt is not None and self.dt > 0:
            raise DREAMException("TimeStepper: 'nt' may not be set alongside 'dt'.")
            
        self.nt = int(nt)


    def setNumberOfSaveSteps(self, nSaveSteps):
        """
        Sets the number of time steps to save to the output file.
        This number must be <= Nt. If 0, all time steps are saved.
        """
        self.nSaveSteps = nSaveSteps


    def setRelTol(self, reltol): self.setRelativeTolerance(reltol=reltol)


    def setRelativeTolerance(self, reltol):
        if reltol <= 0:
            raise DREAMException("TimeStepper: Invalid value assigned to 'reltol': {}".format(reltol))

        self.tolerance.set(reltol=float(reltol))


    def setTerminationFunction(self, func):
        """
        Sets the Python function to call in order to determine when terminate
        the time stepping. **NOTE**: This functionality is only available when
        DREAM is compiled and run as a Python library.

        :param func: Python function determining when to terminate time stepping. Takes a libdreampyface 'Simulation' object as input and returns a bool.
        """
        self.terminatefunc = func


    def setTmax(self, tmax):
        if tmax is None:
            self.tmax = None
            return

        if tmax <= 0:
            raise DREAMException("TimeStepper: Invalid value assigned to 'tmax': {}".format(tmax))

        self.tmax = float(tmax)


    def setType(self, ttype, *args, **kwargs):
        if ttype not in [TYPE_CONSTANT, TYPE_ADAPTIVE, TYPE_IONIZATION]:
            raise DREAMException("TimeStepper: Unrecognized time stepper type specified: {}".format(ttype))

        if ttype in [TYPE_ADAPTIVE, TYPE_IONIZATION]:
            self.nt = None

        self.type = int(ttype)
        
        if ttype == TYPE_IONIZATION:
            self.setIonization(*args, **kwargs)
    

    def setIonization(self, dt0=0, dtmax=0, tmax=None, automaticstep=1e-12, safetyfactor=50):
        """
        Select and set parameters for the ionization time stepper.
        """
        self.type = TYPE_IONIZATION
        self.dt = dt0
        self.dtmax = dtmax
        self.automaticstep = automaticstep
        self.safetyfactor = safetyfactor

        if tmax is not None:
            self.tmax = tmax


    def setVerbose(self, verbose=True):
        self.verbose = bool(verbose)


    def fromdict(self, data):
        """
        Load settings from the given dictionary.
        """
        def scal(v):
            if type(v) == np.ndarray: return v[0]
            else: return v

        self.type = data['type']
        self.tmax = data['tmax']

        if type(self.type) == np.ndarray: self.type = int(self.type.flatten()[0])
        if type(self.tmax) == np.ndarray: self.tmax = float(self.tmax.flatten()[0])

        if 'automaticstep' in data: self.automaticstep = float(scal(data['automaticstep']))
        if 'checkevery' in data: self.checkevery = int(scal(data['checkevery']))
        if 'constantstep' in data: self.constantstep = bool(scal(data['constantstep']))
        if 'dt' in data: self.dt = float(scal(data['dt']))
        if 'dtmax' in data: self.dtmax = float(scal(data['dtmax']))
        if 'minsavedt' in data: self.minsavedt = float(scal(data['minsavedt']))
        if 'nt' in data: self.nt = int(scal(data['nt']))
        if 'nsavesteps' in data: self.nSaveSteps = int(scal(data['nsavesteps']))
        if 'verbose' in data: self.verbose = bool(scal(data['verbose']))
        if 'safetyfactor' in data: self.safetyfactor = float(scal(data['safetyfactor']))
        if 'tolerance' in data: self.tolerance.fromdict(data['tolerance'])
        if 'terminatefunc' in data: self.terminatefunc = data['terminatefunc']
        
        self.verifySettings()


    def todict(self, verify=True):
        """
        Returns a Python dictionary containing all settings of
        this TimeStepper object.
        """
        if verify:
            self.verifySettings()

        data = {
            'type': self.type,
            'tmax': self.tmax
        }

        if self.dt is not None: data['dt'] = self.dt

        if self.type == TYPE_CONSTANT:
            if self.nt is not None: data['nt'] = self.nt
            data['nsavesteps'] = int(self.nSaveSteps)

            if self.terminatefunc != None:
                data['terminatefunc'] = self.terminatefunc
        elif self.type == TYPE_ADAPTIVE:
            data['checkevery'] = self.checkevery
            data['constantstep'] = self.constantstep
            data['tolerance'] = self.tolerance.todict()
            data['verbose'] = self.verbose
        elif self.type == TYPE_IONIZATION:
            if self.dtmax is not None: data['dtmax'] = self.dtmax
            data['automaticstep'] = self.automaticstep
            data['safetyfactor'] = self.safetyfactor
            data['minsavedt'] = self.minsavedt

        return data


    def verifySettings(self):
        """
        Verify that the TimeStepper settings are consistent.
        """
        if self.type == TYPE_CONSTANT:
            if self.tmax is None or self.tmax <= 0:
                raise DREAMException("TimeStepper constant: 'tmax' must be set to a value > 0.")
            
            # Verify that _exactly_ one of 'dt' and 'nt' is
            # set to a valid value
            dtSet = (self.dt is not None and self.dt > 0)
            ntSet = (self.nt is not None and self.nt > 0)

            if dtSet and ntSet:
                raise DREAMException("TimeStepper constant: Exactly one of 'dt' and 'nt' must be > 0.")

            if self.nSaveSteps < 0 or (ntSet and self.nSaveSteps > self.nt):
                raise DREAMException("TimeStepper constant: Invalid value assigned to 'nSaveSteps'. Must between 0 and nt.")
        elif self.type == TYPE_ADAPTIVE:
            if self.tmax is None or self.tmax <= 0:
                raise DREAMException("TimeStepper adaptive: 'tmax' must be set to a value > 0.")
            elif self.nt is not None:
                raise DREAMException("TimeStepper adaptive: 'nt' cannot be used with the adaptive time stepper.")

            if type(self.checkevery) != int or self.checkevery < 0:
                raise DREAMException("TimeStepper adaptive: 'checkevery' must be a non-negative integer.")
            elif type(self.verbose) != bool:
                raise DREAMException("TimeStepper adaptive: 'verbose' must be a boolean.")
            elif type(self.constantstep) != bool:
                raise DREAMException("TimeStepper adaptive: 'constantstep' must be a boolean.")
            self.tolerance.verifySettings()
        elif self.type == TYPE_IONIZATION:
            if self.tmax is None or self.tmax <= 0:
                raise DREAMException("TimeStepper ionization: 'tmax' must be set to a value > 0.")
            elif self.dt is None or self.dt < 0:
                raise DREAMException("TimeStepper ionization: 'dt' must be set to a non-negative value.")
            elif self.dtmax is None or self.dtmax < 0:
                raise DREAMException("TimeStepper ionization: 'dtmax' must be set to a non-negative value.")
            elif self.minsavedt < 0:
                raise DREAMException("TimeStepper ionization: 'minsavedt' must be non-negative.")
        else:
            raise DREAMException("Unrecognized time stepper type selected: {}.".format(self.type))

File: src/Settings/TimeStepper.cpp
/**
 * Construct a time stepper object.
 */

#include "DREAM/EquationSystem.hpp"
#include "DREAM/Settings/SimulationGenerator.hpp"
#include "DREAM/TimeStepper/TimeStepper.hpp"
#include "DREAM/TimeStepper/TimeStepperAdaptive.hpp"
#include "DREAM/TimeStepper/TimeStepperConstant.hpp"
#include "DREAM/TimeStepper/TimeStepperIonization.hpp"
#include "DREAM/TimeStepper/TimeStepperPrescribed.hpp"
#include "FVM/UnknownQuantityHandler.hpp"


using namespace DREAM;
using namespace std;


#define MODULENAME "timestep"


/**
 * Define options for the time stepper.
 *
 * s: Settings object to define settings in.
 */
void SimulationGenerator::DefineOptions_TimeStepper(Settings *s) {
	s->DefineSetting(MODULENAME "/automaticstep", "Step length for the automatic determination of the time step in the ionization time stepper.", (real_t)1e-12);
    s->DefineSetting(MODULENAME "/checkevery", "Check the error every N'th step (0 = check error after _every_ time step)", (int_t)0);
    s->DefineSetting(MODULENAME "/constantstep", "Override the adaptive stepper and force a constant time step (DEBUG OPTION)", (bool)false);
    s->DefineSetting(MODULENAME "/dt", "Length of each time step", (real_t)0.0);
	s->DefineSetting(MODULENAME "/dtmax", "Maximum allowed time step for the adaptive ionization time stepper.", (real_t)0);
	s->DefineSetting(MODULENAME "/minsavedt", "Minimum time required to elapse between time steps to save in adaptive ioniz time stepper.", (real_t)0);
    s->DefineSetting(MODULENAME "/nsavesteps", "Number of time steps to save to output (downsampling)", (int_t)0);
    s->DefineSetting(MODULENAME "/nt", "Number of time steps to take", (int_t)0);
	s->DefineSetting(MODULENAME "/safetyfactor", "Safety factor to use when automatically determining the baseline timestep for the adaptive ionization time stepper.", (real_t)50);
    s->DefineSetting(MODULENAME "/tmax", "Maximum simulation time", (real_t)0.0);
	s->DefineSetting(MODULENAME "/alpha", "Scaling factor for the ionization time. If not zero, timesteps are updated as: dt = min(dtmax, alpha * t_ioniz)", (real_t)0.0);
    s->DefineSetting(MODULENAME "/type", "Time step generator type", (int_t)OptionConstants::TIMESTEPPER_TYPE_CONSTANT);
    s->DefineSetting(MODULENAME "/verbose", "If true, generates excessive output", (bool)false);
    s->DefineSetting(MODULENAME "/times", "Prescribed simulation time points (1D array). The simulation advances from times[k-1] to times[k].", (len_t)0, (const real_t*)nullptr);
#ifdef DREAM_IS_PYTHON_LIBRARY
    s->DefineSetting(MODULENAME "/terminatefunc", "Python function used to determine when to terminate time stepping", (void*)nullptr);
#endif

    // Tolerance settings for adaptive time stepper
    DefineToleranceSettings(MODULENAME, s);
}

/**
 * Construct a TimeStepper object according to the settings.
 *
 * eqsys: Equation system object to assign TimeStepper object to.
 * s:     Settings object specifying how to construct the TimeStepper object.
 */
void SimulationGenerator::ConstructTimeStepper(EquationSystem *eqsys, Settings *s) {
    enum OptionConstants::timestepper_type type = (enum OptionConstants::timestepper_type)s->GetInteger(MODULENAME "/type");

    FVM::UnknownQuantityHandler *u = eqsys->GetUnknownHandler();
    vector<len_t> *nontrivials = eqsys->GetNonTrivialUnknowns();
    TimeStepper *ts;
    switch (type) {
        case OptionConstants::TIMESTEPPER_TYPE_CONSTANT:
            ts = ConstructTimeStepper_constant(s, u, eqsys);
            break;

        case OptionConstants::TIMESTEPPER_TYPE_ADAPTIVE:
            ts = ConstructTimeStepper_adaptive(s, u, eqsys, nontrivials);
            break;

		case OptionConstants::TIMESTEPPER_TYPE_IONIZATION:
			ts = ConstructTimeStepper_ionization(s, u, eqsys);
			break;

        case OptionConstants::TIMESTEPPER_TYPE_PRESCRIBED:
            ts = ConstructTimeStepper_prescribed(s, u, eqsys);
            break;

        default:
            throw SettingsException(
                "Unrecognized time stepper type: %d.", type
            );
    }

#ifdef DREAM_IS_PYTHON_LIBRARY
    void *terminatefunc = s->GetAddress(MODULENAME "/terminatefunc");
    if (terminatefunc != nullptr)
        ts->SetPythonTerminateFunc(terminatefunc);
#endif

    eqsys->SetTimeStepper(ts);
}


/**
 * Construct a TimeStepperConstant object according to the
 * provided settings.
 *
 * s: Settings object specifying how to construct the
 *    TimeStepperConstant object.
 */
TimeStepperConstant *SimulationGenerator::ConstructTimeStepper_constant(
    Settings *s, FVM::UnknownQuantityHandler *u,
    EquationSystem *eqsys
) {
    real_t tmax = s->GetReal(MODULENAME "/tmax");
    real_t dt   = s->GetReal(MODULENAME "/dt", false);
    int_t nt    = s->GetInteger(MODULENAME "/nt", false);
    int_t nSaveSteps = s->GetInteger(MODULENAME "/nsavesteps");

    bool dtset  = (dt > 0);
    bool ntset  = (nt > 0);

    if (dtset && ntset)
        throw SettingsException(
            "TimeStepper constant: "
            "Ambiguous time step specified. Only one of 'dt' and 'nt' "
            "may be set for the time stepper."
        );
    else if (!dtset && !ntset)
        throw SettingsException(
            "TimeStepper constant: "
            "No time step specified. Exactly one of 'dt' and 'nt' "
            "must be set for the time stepper."
        );

    // Generate object
    if (dtset) {
        s->MarkUsed(MODULENAME "/dt");
        return new TimeStepperConstant(tmax, dt, u, eqsys, nSaveSteps);
    } else {
        s->MarkUsed(MODULENAME "/nt");
        return new TimeStepperConstant(tmax, (len_t)nt, u, eqsys, nSaveSteps);
    }
}

/**
 * Construct a TimeStepperAdaptive object according to the
 * provided settings.
 *
 * s: Settings object specifying how to construct the
 *    TimeStepperAdaptive object.
 */
TimeStepperAdaptive *SimulationGenerator::ConstructTimeStepper_adaptive(
    Settings *s, FVM::UnknownQuantityHandler *u,
	EquationSystem *eqsys, vector<len_t> *nontrivials
) {
    int_t checkevery = s->GetInteger(MODULENAME "/checkevery");
    real_t tmax = s->GetReal(MODULENAME "/tmax");
    real_t dt = s->GetReal(MODULENAME "/dt");
    bool verbose = s->GetBool(MODULENAME "/verbose");
    bool conststep = s->GetBool(MODULENAME "/constantstep");

	vector<UnknownQuantityEquation*> *eqns = eqsys->GetEquations();

    if (dt == 0)
        dt = 1;

    ConvergenceChecker *cc = LoadToleranceSettings(
        MODULENAME, s, eqns, u, *nontrivials
    );

    return new TimeStepperAdaptive(tmax, dt, u, eqsys, *nontrivials, eqns, cc, checkevery, verbose, conststep);
}

/**
 * Construct a TimeStepperIonization object according to the
 * provided settings.
 *
 * s: Settings object specifying how to construct the
 *    TimeStepperIonization object.
 */
TimeStepperIonization *SimulationGenerator::ConstructTimeStepper_ionization(
	Settings *s, FVM::UnknownQuantityHandler *u, EquationSystem *eqsys
) {
	real_t automaticstep = s->GetReal(MODULENAME "/automaticstep");
	real_t dt = s->GetReal(MODULENAME "/dt");
	real_t dtmax = s->GetReal(MODULENAME "/dtmax");
	real_t minSaveDt = s->GetReal(MODULENAME "/minsavedt");
	real_t safetyfactor = s->GetReal(MODULENAME "/safetyfactor");
	real_t tmax = s->GetReal(MODULENAME "/tmax");
	real_t alpha = s->GetReal(MODULENAME "/alpha");

	if (dt < 0)
		throw SettingsException("TimeStepper ionization: Initial time step 'dt0' must be non-negative.");

	return new TimeStepperIonization(tmax, dt, dtmax, u, eqsys, automaticstep, safetyfactor, minSaveDt, alpha);
}

/**
 * Construct a TimeStepperPrescribed object according to the
 * provided settings.
 *
 * s: Settings object specifying how to construct the
 *    TimeStepperPrescribed object.
 */
TimeStepperPrescribed *SimulationGenerator::ConstructTimeStepper_prescribed(
    Settings *s, FVM::UnknownQuantityHandler *u,
    EquationSystem *eqsys
) {
    // tmax is still part of the standard interface. For prescribed stepping,
    // we require that (if set) it matches the last time point.
    real_t tmax = s->GetReal(MODULENAME "/tmax");

    // Disallow mixing with dt/nt for clarity.
    real_t dt = s->GetReal(MODULENAME "/dt", false);
    int_t  nt = s->GetInteger(MODULENAME "/nt", false);
    if (dt > 0 || nt > 0)
        throw SettingsException(
            "TimeStepper prescribed: "
            "Ambiguous time step specified. 'dt'/'nt' cannot be used together "
            "with the prescribed time grid 'times'."
        );

    int_t nSaveSteps_i = s->GetInteger(MODULENAME "/nsavesteps");
    if (nSaveSteps_i < 0)
        throw SettingsException(
            "TimeStepper prescribed: "
            "Invalid negative value assigned to 'nsavesteps': %d.", nSaveSteps_i
        );
    len_t nSaveSteps = (len_t)nSaveSteps_i;

    // Load prescribed time array (1D).
    len_t dims[1] = {0};
    const real_t *times = s->GetRealArray(MODULENAME "/times", 1, dims);
    const len_t ntimes = dims[0];

    if (ntimes < 2)
        throw SettingsException(
            "TimeStepper prescribed: "
            "Invalid time grid. 'times' must contain at least two time points."
        );

    // Validate time grid (explicit, no cleverness).
    if (times[0] != 0)
        throw SettingsException(
            "TimeStepper prescribed: "
            "Invalid time grid. The first time point must be 0."
        );

    for (len_t i = 1; i < ntimes; i++) {
        if (times[i] <= times[i-1])
            throw SettingsException(
                "TimeStepper prescribed: "
                "Invalid time grid. 'times' must be strictly increasing."
            );
    }

    const real_t tmax_times = times[ntimes-1];
    if (tmax <= 0) {
        // If user did not set tmax, accept and implicitly use the last time point.
        // (We still keep 'tmax' in settings for consistency with other steppers.)
        tmax = tmax_times;
    } else if (tmax != tmax_times) {
        throw SettingsException(
            "TimeStepper prescribed: "
            "Inconsistent final time. 'tmax' (= %.16e) must match times[end] (= %.16e).",
            tmax, tmax_times
        );
    }

    return new TimeStepperPrescribed(ntimes, times, u, eqsys, nSaveSteps);
}
